### ISSUE TITLE
decoder generates noise images

### ISSUE THREAD

**Issue by xinmiaolin** [lucidrains/DALLE2-pytorch#164]
Hi,

I am trying to train a decoder on CUB-200 dataset. However, after nearly 400 epochs, the decoder can still only generate noises and I have been trying a long time to figure out why. I would appreciate your suggestions!

<img width="729" alt="Screen Shot 2022-06-23 at 10 32 41 AM" src="https://user-images.githubusercontent.com/43487018/175359852-9c4c196f-72f0-4b59-91bc-09b29b13cb16.png">

So the training loss decreases pretty steadily, but the images still look like this:
<img width="794" alt="Screen Shot 2022-06-23 at 10 33 41 AM" src="https://user-images.githubusercontent.com/43487018/175359988-0ed85bfd-9b01-4724-b3f9-9bba67eb3cda.png">

I have only used 1 unet and the configuration is: `dim = 128, image_embed_dim = 768, dim_mults = (1, 2, 4, 8),`, for the decoder, the parameters are: `image_size=64, timesteps = 1000, image_cond_drop_prob = 0.1, text_cond_drop_prob = 0.5, learned_variance = False`.

Thank you!

**Comment by lucidrains** [comment]
@xinmiaolin hi, it looks like your training run diverged early

the loss should go down to around 0.05 before the images come into view

how high is your learning rate?

**Comment by xinmiaolin** [comment]
The learning rate is 4e-3. Yes, the training loss does drop very precipitously from around 1 to 0.1 in the first epoch, then sharply increased to 0.8 again. Then, the training loss decreases steadily. The 0.05 is the loss per image you mean? The batch size is 128 that I used. 

Thank you very much!

**Comment by lucidrains** [comment]
@xinmiaolin do you mean `3e-4` because `4e-3` is absurdly high!

**Comment by lucidrains** [comment]
@xinmiaolin would recommend `1e-4`

**Comment by lucidrains** [comment]
@xinmiaolin yea, the loss should be the MSE, which is average across image samples

**Comment by xinmiaolin** [comment]
I will try a lower learning rate to see if this kind of jumps appear again.

**Comment by lucidrains** [comment]
@xinmiaolin you'll definitely see something

DDPMs are so much easier to train than their predecessor, GANs

**Comment by xinmiaolin** [comment]
> @xinmiaolin you'll definitely see something
> 
> DDPMs are so much easier to train than their predecessor, GANs

Thanks, this definitely gives me hope haha

**Comment by xinmiaolin** [comment]
Hi, I don't think that it is working...

Here is when `lr=1e-4`, although it is diverging, but the loss can still decrease.
<img width="771" alt="Screen Shot 2022-06-24 at 11 28 16 AM" src="https://user-images.githubusercontent.com/43487018/175643665-57016e7e-65fe-4829-926b-d8842d434f5f.png">

This is when `lr=3e-6`, the loss oscillates a lot and does not decrease at all, even increases. 
<img width="771" alt="Screen Shot 2022-06-24 at 11 25 47 AM" src="https://user-images.githubusercontent.com/43487018/175640849-c3cbd9b1-8571-4ec9-be2f-05076749aad4.png">

I am confused why when the lr is smaller, the loss function has such great fluctuations. I am now training with `lr=1e-5`, it looks good. Will keep on updating!

**Comment by xinmiaolin** [comment]
> @xinmiaolin how high is your batch size?

the batch size is 80

**Comment by lucidrains** [comment]
ohh that should be good enough

ok, keep at it with `1e-5`, i'd be surprised if that didn't work

**Comment by xinmiaolin** [comment]
> ohh that should be good enough
> 
> ok, keep at it with `1e-5`, i'd be surprised if that didn't work

ok, I will come back and update. Thanks

**Comment by lucidrains** [comment]
@xinmiaolin i'll add a learning rate warmup for the decoder trainer some time this weekend; just need to figure out how to make it huggingface accelerate compatible

**Comment by xinmiaolin** [comment]
> @xinmiaolin i'll add a learning rate warmup for the decoder trainer some time this weekend; just need to figure out how to make it huggingface accelerate compatible

Ok thanks!

**Comment by xinmiaolin** [comment]
This is the training of `lr=1e-5`, I have also used CosineAnnealingLR with `t_max=10000`. The loss increases from around 0.3 to 0.6 then does not decrease at all.

<img width="1089" alt="Screen Shot 2022-06-26 at 10 59 54 AM" src="https://user-images.githubusercontent.com/43487018/175827698-0d4926b7-a55b-41f5-859a-44cdbb50dd27.png">

For `lr=3e-5`, the loss increases from around 0.2 to 0.7, then decreases slowly.
<img width="1089" alt="Screen Shot 2022-06-26 at 11 04 27 AM" src="https://user-images.githubusercontent.com/43487018/175827874-147bb97b-366c-4412-a3b5-6cc648e8cc84.png">

I have tried smaller learning rates without the lr scheduler, but the loss also fluctuates, for example, when `lr=3e-7`. 

<img width="764" alt="Screen Shot 2022-06-26 at 11 07 45 AM" src="https://user-images.githubusercontent.com/43487018/175828020-5c67ad8a-63b6-4986-b9ea-ff4bade08000.png">

I am not sure what is going on. Could there be some parameters related to diffusion models that should be changed, because I am using default values of the parameters.

**Comment by lucidrains** [comment]
@xinmiaolin hmm, could you possibly send me your full training script?

**Comment by xinmiaolin** [comment]
> @xinmiaolin hmm, could you possibly send me your full training script?

sure, thanks!

**Comment by YUHANG-Ma** [comment]
Did you check your optimizer? I think it maybe because your parameters are not doing the backward. I met the problem before, the loss decreased  slowly but I figured out there is something wrong in optimizer.py.

**Comment by xinmiaolin** [comment]
> Did you check your optimizer? I think it maybe because your parameters are not doing the backward. I met the problem before, the loss decreased slowly but I figured out there is something wrong in optimizer.py.

Hi thank you for the suggestion. I will check on it.

**Comment by BBaDBoYY** [comment]
In the sample code, the input for the text is a random tensor, how should I get this tensor from a piece of text for a given dataset?
`text = torch.randint(0, 49408, (4, 256)).cuda()`
@lucidrains

**Comment by BBaDBoYY** [comment]
@xinmiaolin Have you solved the problem or have you trained the model yet, I have the same problem

**Comment by thecooltechguy** [comment]
Facing the same problem, would love to know if you found out the issue! :)

cc: @xinmiaolin

**Comment by canelle20214** [comment]
@xinmiaolin Same for me, did you found out the root cause of that behaviour ?

**Comment by xiaomei2002** [comment]
Did you figure it out?? I met the problem too.
